collectors/build/jenkins/src/main/java/com/capitalone/dashboard/collector/DefaultHudsonClient.java

```diff
@@ -358,14 +360,42 @@ private BuildStatus getBuildStatus(JSONObject buildJson) {
                 return BuildStatus.Unknown;
         }
     }
-
-    protected ResponseEntity<String> makeRestCall(String sUrl) throws MalformedURLException {
+    
+    @SuppressWarnings("PMD")
+    protected ResponseEntity<String> makeRestCall(String sUrl) throws MalformedURLException, URISyntaxException {
         URI thisuri = URI.create(sUrl);
         String userInfo = thisuri.getUserInfo();
 
         //get userinfo from URI or settings (in spring properties)
-        if (StringUtils.isEmpty(userInfo) && (this.settings.getUsername() != null) && (this.settings.getApiKey() != null)) {
-            userInfo = this.settings.getUsername() + ":" + this.settings.getApiKey();
+        if (StringUtils.isEmpty(userInfo)) {
+        	List<String> servers = this.settings.getServers();
+        	List<String> usernames = this.settings.getUsernames();
+        	List<String> apiKeys = this.settings.getApiKeys();
+        	if (servers != null && usernames != null && apiKeys != null) {
+        		boolean exactMatchFound = false;
+	        	for (int i = 0; i < servers.size(); i++) {
+	        		if ((servers.get(i) != null)) {
+	        			String domain1 = getDomain(sUrl);
+	        			String domain2 = getDomain(servers.get(i));
+	        			if (domain1 != null && domain2 != null && domain1.equals(domain2)
+	        					&& getPort(sUrl) == getPort(servers.get(i))) {
+	                		exactMatchFound = true;	
+	        			}
+	        			if (exactMatchFound && (i < usernames.size()) && (i < apiKeys.size()) 
+	        					&& (usernames.get(i) != null) && (apiKeys.get(i) != null)) {
+	        				userInfo = usernames.get(i) + ":" + apiKeys.get(i);
+        				}
+	        			if (exactMatchFound) {
+	        				break;
+	        			}
+	        		}
+	        	}	        	
+        		if (!exactMatchFound) {
+        			LOG.warn("Credentials for the following url was not found. This could happen if the domain/subdomain/IP address "
+        					+ "in the build url returned by Jenkins and the Jenkins instance url in your Hygieia configuration do not match: "
+        					+ "\"" + sUrl + "\"");
```

Comment:
Yes, as you know, the build URL is read from the Jenkins JSON output and that is used to make REST calls and get the build details. So I don't see how to avoid a reverse lookup to find the appropriate credentials. And in case of redirects, should we just force the user to enter the redirect address in the properties file?